src/visitors/sympy_replace_solutions_visitor.cpp

```diff
@@ -0,0 +1,350 @@
+    is_statement_block_root_ = false;
+
+    if (is_root) {
+        logger->debug("SympyReplaceSolutionsVisitor :: visit statements. Matching policy: {}",
+                      (policy_ == ReplacePolicy::VALUE ? "VALUE" : "GREEDY"));
+        interleaves_counter_ = InterleavesCounter();
+        node.visit_children(*this);
+
+        if (!solutions_.tags_.empty() && policy_ == ReplacePolicy::VALUE) {
+            logger->debug(
+                "SympyReplaceSolutionsVisitor :: not all solutions were replaced. Policy: GREEDY");
+            interleaves_counter_ = InterleavesCounter();
+            policy_ = ReplacePolicy::GREEDY;
+            node.visit_children(*this);
+
+            if (interleaves_counter_.n() > 0) {
+                logger->warn(
+                    "SympyReplaceSolutionsVisitor :: Found ambiguous system of equations "
+                    "interleaved with {} assignment statements. I do not know what equations go "
+                    "before and what "
+                    "equations go after the assignment statements. Either put all the equations "
+                    "that need to be solved "
+                    "in the form: x = f(...) and with distinct variable assignments or do not "
+                    "interleave the system with assignments.",
+                    interleaves_counter_.n());
+            }
+        }
+    } else {
+        node.visit_children(*this);
+    }
+
+    const auto& old_statements = node.get_statements();
+
+    ast::StatementVector new_statements;
+    new_statements.reserve(2 * old_statements.size());
+    for (const auto& old_statement: old_statements) {
+        const auto& replacement_ptr = replacements_.find(old_statement);
+        if (replacement_ptr != replacements_.end()) {
+            if (replaced_statements_begin_ == -1) {
+                replaced_statements_begin_ = new_statements.size();
+            }
+
+            new_statements.insert(new_statements.end(),
+                                  replacement_ptr->second.begin(),
+                                  replacement_ptr->second.end());
+
+            replaced_statements_end_ = new_statements.size();
+
+            logger->debug("SympyReplaceSolutionsVisitor :: erasing {}", to_nmodl(old_statement));
+            for (const auto& replacement: replacement_ptr->second) {
+                logger->debug("SympyReplaceSolutionsVisitor :: adding {}", to_nmodl(replacement));
+            }
+        } else if (to_be_removed_ == nullptr ||
+                   to_be_removed_->find(&(*old_statement)) == to_be_removed_->end()) {
+            new_statements.emplace_back(std::move(old_statement));
+        } else {
+            logger->debug("SympyReplaceSolutionsVisitor :: erasing {}", to_nmodl(old_statement));
+        }
+    }
+
+    if (is_root) {
+        if (!solutions_.tags_.empty()) {
+            throw std::runtime_error(
+                "Not all solutions were replaced! Probably something went wrong.");
+        }
```

Comment:
not very easy since this error is only for developers. If the user reaches this point there was a bug somewhere before. In any case, I improved it because utimately, yes, I agree it was a shitty error message